Fail job if job output collection fails

mvdbeek opened the pull request:
We now set a non-zero exit code to actually set a job to failed.
I believe this is a reasonable action, and it avoids green, empty datasets
with log messages like:
```
galaxy.jobs.runners ERROR 2017-11-14 13:32:06,469 (19020/860246.torque6.curie.fr) Job output not returned from cluster: [Errno 2] No such file or directory: '/data/users/mvandenb/gx
124/tmp_nfs/jwd/019/19020/galaxy_19020.o'
galaxy.jobs.runners DEBUG 2017-11-14 13:32:06,491 (19020/860246.torque6.curie.fr) Unable to cleanup /data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.o: [Errno 2] No suc
h file or directory: '/data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.o'
galaxy.jobs.runners DEBUG 2017-11-14 13:32:06,503 (19020/860246.torque6.curie.fr) Unable to cleanup /data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.e: [Errno 2] No suc
h file or directory: '/data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.e'
galaxy.jobs.runners DEBUG 2017-11-14 13:32:06,513 (19020/860246.torque6.curie.fr) Unable to cleanup /data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.ec: [Errno 2] No su
ch file or directory: '/data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.ec'
```

mvdbeek committed:
Fail job if job output collection fails

We now set a non-zero exit code to actually set a job to failed.
I belive this is a reasonable action, and it avoids green, empty datasets
with log messages like:
```
galaxy.jobs.runners ERROR 2017-11-14 13:32:06,469 (19020/860246.torque6.curie.fr) Job output not returned from cluster: [Errno 2] No such file or directory: '/data/users/mvandenb/gx
124/tmp_nfs/jwd/019/19020/galaxy_19020.o'
galaxy.jobs.runners DEBUG 2017-11-14 13:32:06,491 (19020/860246.torque6.curie.fr) Unable to cleanup /data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.o: [Errno 2] No suc
h file or directory: '/data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.o'
galaxy.jobs.runners DEBUG 2017-11-14 13:32:06,503 (19020/860246.torque6.curie.fr) Unable to cleanup /data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.e: [Errno 2] No suc
h file or directory: '/data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.e'
galaxy.jobs.runners DEBUG 2017-11-14 13:32:06,513 (19020/860246.torque6.curie.fr) Unable to cleanup /data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.ec: [Errno 2] No su
ch file or directory: '/data/users/mvandenb/gx124/tmp_nfs/jwd/019/19020/galaxy_19020.ec'
```

nsoranzo commented on lib/galaxy/jobs/runners/__init__.py:
```diff
@@ -623,13 +623,14 @@ def finish_job(self, job_state):
         try:
             # This should be an 8-bit exit code, but read ahead anyway:
-            exit_code_str = open(job_state.exit_code_file, "r").read(32)
+            exit_code_str = open(job_state.exit_code_file, "r").read(32) if not exit_code_str else exit_code_str
```
This works properly only if `exit_code_str` is always defined, otherwise this would result in:
```
NameError: name 'exit_code_str' is not defined
```
which is caught by the next `except`, ending up almost always with a value of `"0"`.

mvdbeek replied:
uff, of course. thanks for the catch.

mvdbeek committed:
Fix undefined exit_code_str problem (thx @nsoranzo)

jmchilton commented:
At first glance I don't love this I don't think. This wouldn't cause the datasets to fail if the exit code isn't used a condition to fail the job - which is the old default and a reasonable way to configure even newer tools. I feel like we should somehow more directly fail the job. We could either always fail the job when this exit code is encountered else where (maybe in job finish?) or better yet somehow set the exit code to ``None`` and fail all tool-based jobs with ``None`` exit codes?

mvdbeek committed:
Mark job wrapper as failed in a more direct way

mvdbeek commented:
Alright, I have done some manual testing, and this properly fails jobs for tools that would previously end up as OK.
![screen shot 2017-11-28 at 12 19 08](https://user-images.githubusercontent.com/6804901/33317163-ad345d86-d436-11e7-928c-57b0262558ab.png)

mvdbeek committed:
Add new runner state for 'Job output not returned from cluster'

This would allow adimns to write resubmit conditions for this specific failure.

mvdbeek commented:
With the last commit admins could even setup job resubmit conditions for this specific failure.

jmchilton commented:
Thanks so much for the contribution @mvdbeek - this is awesome.

mvdbeek commented:
@erasche that may be of interest to the failures you were seeing
